### ISSUE TITLE
Misc: To release 1.0.25

### ISSUE THREAD

**Issue by mcuee** [libusb/libusb#777]
Merged main feature improvement
- #911
- #935
- #312
- #830

Merged main bug fixes
- #876 
- #852
- #825 
- #850 
- #858 
- #913
- #912
- #863
- #844
- #1016 (fix libusb0 regression in 1.0.24)
- #986
- #967
- #963
- #1010 (workaround for broken device)
- #971
- #996
- <s>#989</s> (unreleased regression)
- <s>#995</s> (unreleased regression in error path)
- <s>#942</s> (avoid regression on Android)
- <s>#1031</s> (unreleased breakage on new macOS 12)
----

Postponed for 1.0.26 (see #1047 )
- #1003 (possible API breakage)
- #1026 (solve API issue)
- #1018 (libusb0.sys)
- #678 (postpone unless reproduced issue)
- #965 (very useful feature enhancement, especially for Windows users, new APIs)

**Comment by mcuee** [comment]
With #838 and #831 and a few other issues resolved, probably it is good to prepare for a maintainence release 1.0.25.

**Comment by mcuee** [comment]
I think #850 will be fixed soon as well.

**Comment by mcuee** [comment]
Pull request $833 should be in as well.
https://github.com/libusb/libusb/pull/833

**Comment by yorickvP** [comment]
Hi! Great work fixing these issues! I just ran into #831 and spent a lot of time debugging until someone pointed me to the issue. It might be worth it to release 1.0.25 so nobody ends up using 1.0.24 by accident (it's the default in some distros).

**Comment by mcuee** [comment]
I think the following will be good to be included in the 1.0.25 release:
#833 (pending review), #856 (reviewed, pending minor modification of commit message) and #850 (pending review of #851).

**Comment by tormodvolden** [comment]
Please consider #847 (pending review of #852) as well, it appears as a regression.

**Comment by mcuee** [comment]
Thanks to Chris. #833 and #850 have been closed.

**Comment by mcuee** [comment]
@hjelmn Please help to review #852 and then see if you want to consider that for the release. Thanks.

**Comment by EmptyVisual** [comment]
@mcuee Can we do a partial maintenance release of the Linux regressions in #856 and #858 whilst we wait for the Darwin issues to be reviewed and merged separately?  Most of us Linux users have been forcibly sitting on 1.0.23.0 for months now because of the bugs in the 1.0.24.0 release.

**Comment by mcuee** [comment]
I think #856 needs to be closed for the release to happen.

**Comment by tormodvolden** [comment]
+1 for high time to get a maintenance release with regression fixes out. I asked about #852 (Darwin) here two months ago because it is relatively trivial, but this is not a regression fix so please don't let it delay any release. Indeed it can seem like the release is hanging on #856, where the suggested fix is a massive commit with a mix of code refactoring and actual changes, API changes even. IMO it is not a critical bug fix (it was discussed whether it was a bug at all) so it could wait until after 1.0.25 and bake longer in the tree before going into a release.

**Comment by mcuee** [comment]
#856 fixed #858. So I think #856 is still good to be merged. Nathan already reviewed it and he mentioned just needs to change the commit message. So I think it is better to merge it before the release.

**Comment by mcuee** [comment]
@dickens Please weigh in as well. Thanks.

**Comment by mcuee** [comment]
As far as I am concerned, only #856 needs to be merged and we should be ready for an RC.

**Comment by mcuee** [comment]
#856 has been merged.

**Comment by mcuee** [comment]
BTW, welcome to join the maintainer team. Right now only Nathan and Chris are actively handling the release.

Release since 2015.
libusb 1.0.24: by Chris on 11 Dec 2020
libusb 1.0.23 by Nathan on on 29 Aug 2019
libusb 1.0.22 by Chris on 25 Mar 2018
libusb 1.0.21 by Nathan on 26 Oct 2016
libusb 1.0.20 by Chris on 14 Sep 2015

**Comment by mcuee** [comment]
The main difficulty with libusb is probably on Windows backends. As of today, there are 48 issues and half of them are related to Windows. And one of the main missing feature under Windows is Hotplug. 

But never mind if you do not want to handle Windows, you can still help to handle Linux and/or macOS.

**Comment by mcuee** [comment]
For those who are interested to see the 1.0.25 release coming soon, please help to carry out tests on git head. Thanks.

**Comment by hjelmn** [comment]
I identify why the Darwin backends is failing the stress test. Will have a fix today.

**Comment by hjelmn** [comment]
Identified the issue with Linux and the stress test. Fixing.

**Comment by hjelmn** [comment]
Interesting question. I have always assumed this is true but we don't document it. We can call:

```
libusb_set_option(NULL, ...);
```

before the default context is created, right? If not the stress test is wrong. Otherwise I have a fix that will allow this. Will hold off of this change until I get a consensus.

**Comment by hjelmn** [comment]
Opened #923 as a prototype. Please comment.

**Comment by mcuee** [comment]
#923 seems to be good to me.

**Comment by mcuee** [comment]
> I identifed why the Darwin backend is failing the stress test. Will have a fix today.
> 
> The problem is in the ordering in libusb_exit caused by destroying the cached device list when the init count goes to 0. Still not sure I agree with this behavior change. I used a library destructor function for a reason... I can fix the ordering issue but this code is somewhat more fragile than I would like.

Commit 676076c helps macOS stress test to have the same result as Linux and Windows. Then #923 should be able to fix the remaining segfault issues with Stress test.

**Comment by patstew** [comment]
Any chance of getting #867 in? And my #312 too if there aren't any objections.

**Comment by mcuee** [comment]
@patstew I am fine with #867. But #312 will need review from Chris.

**Comment by hjelmn** [comment]
Same here. Do we know if Chris has cycles to take a look?

**Comment by mcuee** [comment]
@hjelmn Looks like Chris is busy. We can take the risk to merge #312 if we do not hear from Chris. I think the risk is not high.

**Comment by hjelmn** [comment]
I think we can take a risk on #312. As I mentioned in that PR it is a pretty big missing feature in the backend.

**Comment by mcuee** [comment]
@hjelmn I agree with you. Let's merge #312 as well.

**Comment by hjelmn** [comment]
Done.

**Comment by hjelmn** [comment]
I can't comment at all on #867 but it seems safe. We could include it and see if any issues come up in the RC. I don't know how much testing it will get though.

**Comment by mcuee** [comment]
#867 should be safe.

**Comment by hjelmn** [comment]
Done. Will merge my PR as well then I think we are good to go.

**Comment by hjelmn** [comment]
Ok, ready to test again.

**Comment by mcuee** [comment]
Great. I will test the current git head again this evening.

**Comment by mcuee** [comment]
I tend to think #924 may not be a big issue and should not prevent RC1. So we are probably good to go for RC1 now.

**Comment by mcuee** [comment]
But if we want to take a bit more risks, I am also fine with #895, #900 and #925. Tormod Volden has reviewed #895 and he is the author of #900 and #925. I think he knows libusb pretty well and the fixes all seem to be reasonable to take.

**Comment by mcuee** [comment]
@hjelmn The other common complaints about libusb is for Android which I can not test myself due to lack of knowledge and equipment. But I think #830 and #874 are in the right direction to improve Android. #874 is apparently not ready to be merged. As for #830 I think we at least can merge the improved README file for Android which does not hurt anything.

**Comment by mcuee** [comment]
As for #912, I found it to be there already in 1.0.24 and libusbk project is now considered unmaintained, so I am okay with not fixing it in 1.0.25. We will encourage people to use WinUSB whenever possible.

**Comment by mcuee** [comment]
Looks like #925 needs some more discussions.

**Comment by tormodvolden** [comment]
- #896 I commented on it now. It sounds fine, but the code details are quite hard to review. I am sure the author understands it well and tested it though.
- #900 Would be nice to have @dickens' blessing, but I think it is a safe improvement.
- #925 The pull request is safe. The question is is whether to already introduce a new option name, and obsolete the old. This just get worse the longer we leave the old option name in and supported.
- #852 Is also ready from my side, @hjelmn MIA? Also a safe improvement IMO.

It seems like the prospect of an RC is bringing up activity on a lot of long-time, low-hanging fruit, and we should profit from the momentum to get this through. With the fire-fighting after #856 we should probably let it cool down a couple weeks before final release though.

**Comment by mcuee** [comment]
@tormodvolden Thanks a lot for the help. I am fine with #852 as well.

And yes, after RC1 we will sure have two weeks or more cool down time before release.

**Comment by mcuee** [comment]
@hjelmn @tormodvolden I have narrowed down #924 on external hub. Hopefully you can take a look and find something from the run logs.

**Comment by hjelmn** [comment]
@mcuee I will take a look. I have a couple of external hubs I can test with. Will be later today or in the evening.

**Comment by mcuee** [comment]
@hjelmn Looks like #924 may already be present in 1.0.24 and may not be a common issue. So I think it is okay to proceed with RC1 without resolving it.

**Comment by mcuee** [comment]
@hjelmn @tormodvolden Please take a look at #912 to see if the test code is valid or not. It is related to Windows and libusbk driver, but you do not need to care for Windows, just to see if the example code is valid or not.

**Comment by mcuee** [comment]
With regard to #830, I think it is safe to include the updated README file but not the example.

**Comment by mcuee** [comment]
BTW, I have done some simple tests under Linux, macOS and Windows, it seems to work fine other than the issues raised here.

I also tested using libusbdotnet Benchmarkcon example under Linux, macOS and Windows, it indirectly test libusb as well and it works fine.

**Comment by hjelmn** [comment]
#924 should be fixed now.

**Comment by hjelmn** [comment]
Please test and ensure I didn't break anything.

**Comment by mcuee** [comment]
#900 and #852 have been merged. I will test the latest git under macOS and Linux. It seems to work fine under Windows 10.

**Comment by mcuee** [comment]
So remaining question is #896 review and #925 improvement.

**Comment by mcuee** [comment]
Somehow #924 becomes a different issue now with the latest git.

**Comment by mcuee** [comment]
#929 can be merged as well. Somehow it was missed in 1.0.23 release due to a misunderstanding.

**Comment by mcuee** [comment]
#924 is now resolved.

**Comment by mcuee** [comment]
Now only #929 and #927 and #935 need to be merged.

**Comment by mcuee** [comment]
@hjelmn Just wondering if you have got the time to merge the above and then create RC1. Thanks.

**Comment by mcuee** [comment]
I have done further testing of #912 and it seems to be a bit complicated. Since it is already in 1.0.24 and only concerns libusbk and limited test cases, I will say just live with it unless Chris or others got the time to look into it.

**Comment by hjelmn** [comment]
Been a bit busy the last couple of weeks. Will take a look today.

**Comment by mcuee** [comment]
#943 is also good to have since it fixed #912.
@tormodvolden Please take a look at #943 if you got the time. Thanks.
@dickens Please take a look at #943 as well in case you have got some time. Thanks.

**Comment by mcuee** [comment]
@hjelmn Under macOS #945 is more serious than #930 so it may be worth looking at as well.

**Comment by mcuee** [comment]
@hjelmn For macOS, #929 (fixes #410) is a simple fix to be merged. #930 may only require a simple change as well. #945 may be more complicated.

Not so sure about #678 and #908.

**Comment by mcuee** [comment]
@hjelmn @LudovicRousseau For Linux/Android, maybe #935 is good to go. And probably also #942 is worth looking at.

**Comment by hjelmn** [comment]
#929 is done. Though at some point we should cut off 10.7 entirely. I have no interest in supporting ancient OS versions :)

**Comment by hjelmn** [comment]
Merged #927, #935, and #943. I will look at the macOS issue but at this point regression fixes only for the next release.

**Comment by mcuee** [comment]
I agree we should stop supporting ancient macOS, probably next release (1.0.26). I will test out latest git against #945 soon. Then we should be good for RC1 already.

**Comment by mcuee** [comment]
@hjelmn Thanks. 

Now that #945 and #930 have been fixed, for macOS, probably only #908 is worth a look. 

For Windows and Linux (not Android) I do not see anything already. 

For Android, I am not so sure if #942 is still necessary.

**Comment by mcuee** [comment]
@hjelmn Or probably we should just go with 1.0.25RC1 now. I have done some testing on the macOS side using the built-in examples and it seems to be fine now.

**Comment by mcuee** [comment]
@hjelmn Two new things come in.
1) Windows: https://github.com/libusb/libusb/pull/948 can be merged.
2) macOS: please review https://github.com/libusb/libusb/pull/950

**Comment by mcuee** [comment]
@hjelmn Great, most of the things have been merged. I understand that you will take a look at #942.

**Comment by mcuee** [comment]
@hjelmn Other than #942, is it possible to merge #971 for Android? I think it is a safe fix affecting only Android.

**Comment by xloem** [comment]
Is #942 good to close out for this?  It's very simple, and it would be nice to restabilise the in-tree android code.  I'm not sure whether old code for non-rooted android currently runs without it.

**Comment by tormodvolden** [comment]
@mcuee Just a suggestion: What about maintaining a list of issues in your first post, something like (I might have missed or misplaced some!):

Ready to merge:
- #986
- #971

Pending review:
- #942

Targeted:
- #678 
- #908
- #963

Under consideration:

Postponed:
- #965

**Comment by mcuee** [comment]
@tormodvolden Thanks for the suggestion. 

Your list seems to be pretty good. Only two minor changes from my side.
1) #908 has been closed. 
2) I put #965 as under consideration. Let's see whether Nathan is comfortable with the changes or not.

**Comment by tormodvolden** [comment]
You could add #830 (changes requested) to "under consideration". Also #678 has changes requested. It is a question how much information to add to this list, but comments as "feature" and "regression" can be useful to have a good overview. We could also add a section for fixed issues (regressions, normal bugs, features) and move things there as we go, which could help writing release notes.

What I don't like about the current process is that we had several regressions fixed last year, but the release is still being kept back by adding new stuff and sometimes bug fixes that turns out more complicated than expected. We shouldn't wait for all known bugs to be fixed but quickly release regression fixes.

**Comment by mcuee** [comment]
I have updated the list. And add the Feature and Main bug fixes list.

**Comment by tormodvolden** [comment]
We should add #989 to the pending review list.

**Comment by mcuee** [comment]
Tormod has joined the libusb team and he has merged quite a few pull-requests. Post 1 updated.

**Comment by xloem** [comment]
Should #942 be a blocker for this release, so the release doesn't break existing rootless android code?  Or maybe hjelmn's changes to option handling that stimulated 942 could be pulled out into a PR?

**Comment by Tpimp** [comment]
Anyone following pull request #965  should know I have additional changes I'm working to get in with this PR.
[Add dynamic DeviceInterfaceGUIDs support](https://github.com/jdkvision/libusb/pull/2)

Hopefully this can all land in 1.0.25 as I am seeing significant improvements in my application on Windows using the changes from Ryan. We are still hoping for more review and feedback from anyone that has time/ability to test on Windows with these changes.

**Comment by mcuee** [comment]
> Anyone following pull request #965 should know I have additional changes I'm working to get in with this PR. [Add dynamic DeviceInterfaceGUIDs support](https://github.com/jdkvision/libusb/pull/2)
> 
> Hopefully this can all land in 1.0.25 as I am seeing significant improvements in my application on Windows using the changes from Ryan. We are still hoping for more review and feedback from anyone that has time/ability to test on Windows with these changes.

Can you create a pull request here in libusb github repo? In that case, this will not be forgotten. 

Personally I would really like to get this merged in 1.0.25 release, unfortunately we are lacking Windows side developers as of now.

**Comment by hjelmn** [comment]
I think other than #1011 we are ready for a release. I really want to get that one solved. That said, it may take some time to debug since it looks like I need to get a different device to trigger the bug so maybe cut a new RC and give me a week. If it is not resolved by then we can go ahead and release.

**Comment by tormodvolden** [comment]
@hjemn, did you look at the list in the first post here? IMO we absolutely must consider the two under "pending review", one is even a regression since 1.0.24, and the one under "must be checked" other than #1011.

**Comment by hjelmn** [comment]
@tormodvolden Ah, didn't see the list was updated. Looking.

**Comment by mcuee** [comment]
> @hjelmn, did you look at the list in the first post here? IMO we absolutely must consider the two under "pending review", one is even a regression since 1.0.24, and the one under "must be checked" other than #1011.

Move the "must be checked" in front of "pending review".

**Comment by mcuee** [comment]
So that the following two are pending for the RC release.
- #1003 
- #1026

**Comment by hjelmn** [comment]
Yeah. I can merge #1026 but only if I can get one more review. Adding an API should be done with care and I don't want to have to replace ```libusb_init()``` more than once.

**Comment by hjelmn** [comment]
Once that is resolved we need to cut a new RC and get things rolling. We solved the single biggest issue blocking macOS.

**Comment by mcuee** [comment]
> So that the following two are pending for the RC release.
> 
> * [no_enumeration should be per context, not global #1003](https://github.com/libusb/libusb/issues/1003)
> * [core: add libusb_init_context to allow setting options during context creation #1026](https://github.com/libusb/libusb/pull/1026)

Just wondering if there are any updates on these two last remaining main issues?

**Comment by tormodvolden** [comment]
IMO, if they are not fixing regressions, these should wait for next release, and we should make the RC (and possibly the 1.0.25 release) with what we have in git now. I previously wanted these two items included, but as we discussed in #1026, these major API changes can need more review and no rushing. And we dearly need a 1.0.25 now.

**Comment by mcuee** [comment]
I agree. We can probably tag the release now, or at least the RC1.

**Comment by mcuee** [comment]
@hjelmn @LudovicRousseau What are your opinions?

**Comment by LudovicRousseau** [comment]
No objection for a release soon.

"Release early, release often". Even if we do not have the manpower for the "release often".

**Comment by hjelmn** [comment]
Go for it!

**Comment by tormodvolden** [comment]
RC1 is out, please help fill in (and extend) the test matrix in #1046. Thanks Ludovic for the autobuild stuff, that helps already build testing on a load of Windows platforms.

**Comment by mcuee** [comment]
@tormodvolden @hjelmn @LudovicRousseau 
It seems that we are more or less ready for the 1.0.25 release. Anything else we would like to fix?

**Comment by hjelmn** [comment]
Looks good to me.

**Comment by tormodvolden** [comment]
Is there anything missing in the ChangeLog or other things worth mentioning in the release announcement?

Does anyone have the Windows binaries package *.7z build automatized or can volunteer to create it once we have the release tarball?

**Comment by mcuee** [comment]
> Is there anything missing in the ChangeLog or other things worth mentioning in the release announcement?
> 
> Does anyone have the Windows binaries package *.7z build automatized or can volunteer to create it once we have the release tarball?

I think the changelog is okay.

As for the Windows binary, it is okay not to have that initially. I remember that was the case when we released 1.0.23. I actually do not know how the 7z Windows binary packge is created. It is not a real issue though.
1) Users with Visual Studio can easily build from source
2) vcpkg has already libusb package
3) MSYS2 distro will also have libusb package as well.

**Comment by mcuee** [comment]
The MinGW binary package is built with the following script, but you can see it is using a multilib enabled toolchain. I am not so sure what is a suitable toolchain to use in this case.
https://github.com/libusb/libusb/blob/master/.private/bm.sh

**Comment by tormodvolden** [comment]
The 1.0.24 .7z has a long list of VS targets that correspond exactly to the Appveyor builds, so I wonder if Appveyor was used to build them, maybe locally. @LudovicRousseau's web service instance doesn't seem to store build artifacts.

**Comment by mcuee** [comment]
I think we can tag 1.0.25 release first so that various distributions (Linux distro, macOS Homebrew and macports, Windows vcpkg and Msys2) can take the new releases. 

We can always release the official windows binary once we figured out the way. If Chris is not available, probably Pete can help as well.

**Comment by LudovicRousseau** [comment]
@tormodvolden I have not configured github actions to store any artifacts. That was on purpose. I don't think we need them for Unix.
It is possible to generate and store artifacts for Windows builds. But I will not take care of Windows builds myself.

Maybe the lack of Windows builds will motivate some Windows developers to help the project? :-)

**Comment by tormodvolden** [comment]
@LudovicRousseau I was thinking of your Appveyor builds for Windows. I think storage of artifacts can easily be enabled there, but I fully understand if you don't want to get involved in it.

**Comment by LudovicRousseau** [comment]
@tormodvolden I completely forgot Appveyor :-)
Everything was configured and developed by @dickens, not me.

**Comment by tormodvolden** [comment]
I have configured appveyor so it stores build artifacts. So I can spin a .7z with all VS builds plus MinGW cross-build, and have it uploaded before I announce the release. Tentatively on Monday night (UTC).

**Comment by mcuee** [comment]
That will be great. Thanks.

**Comment by mcuee** [comment]
Wonderful that we finally have the libusb 1.0.25 release.

BTW, here in Singapore it is Chinese New Year holiday today.

**Comment by tormodvolden** [comment]
That was good timing :) Happy New Year!

If the release is as quiet as the RC, we're good. It is already picked up by e.g. homebrew.
